### PR TITLE
Feature/pelagos 3939 create forms and ui

### PR DESCRIPTION
**Pull request by mickel1138**

### CODE REVIEW

**Comment by mickel1138** [comment]
Needs to have unit tests fixed.

File: src/Entity/DatasetLink.php [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+    }
+
+    /**
+     * Set the function code for the Dataset Link.
+     *
+     * @param string $functionCode The function code for the Dataset Link.
+     *
+     * @return self
+     */
+    public function setFunctionCode(?string $functionCode): self
```
Comment (praneethpr) [inline src/Entity/DatasetLink.php]: Same here

File: src/Entity/DatasetLink.php [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+    }
+
+    /**
+     * Set the dataset submission for this link.
+     *
+     * @param DatasetSubmission|null $datasetSubmission The dataset submission for this link.
+     *
+     * @return self
+     */
+    public function setDatasetSubmission(?DatasetSubmission $datasetSubmission): self
```
[the remark this answers is not in the document]
Comment (mickel1138) [reply src/Entity/DatasetLink.php]: This is what the maker bundle did. I questioned it myself at first too, but I think it's so you can do things like this:

`$myObject = $myObject->changeItem($item);`

I guess that can be useful is certain situations, but if you think we should do things they same everywhere, then that's ok too. Writing the test is harder now.

[the remark this answers is not in the document]
Comment (mickel1138) [reply src/Entity/DatasetLink.php]: I think this also help with New-ing an Class.

File: src/Entity/DatasetSubmission.php [hunk 3]
```diff
@@ -1155,6 +1166,18 @@ public function __construct(Entity $entity, PersonDatasetSubmissionDatasetContac
                 $newDistributionPoint->setDataCenter($distributionPoint->getDataCenter());
                 $this->addDistributionPoint($newDistributionPoint);
             }
+
+            // Copy the original Dataset Submission's Dataset Links.
+            foreach ($entity->getDatasetLinks() as $datasetLink) {
+                $newDatasetLink = new DatasetLink();
+                $newDatasetLink->setUrl($datasetLink->getUrl());
```
Comment (praneethpr) [inline src/Entity/DatasetSubmission.php]: If we are going to use return types for setters, then we should use method chaining as well! I think if aren't going to use method chaining, we shouldn't use return type for setters.

For example:

`class fakeString
{
    private $str;
    function __construct()
    {
        $this->str = "";
    }

    function addA()
    {
        $this->str .= "a";
        return $this;
    }

    function addB()
    {
        $this->str .= "b";
        return $this;
    }

    function getStr()
    {
        return $this->str;
    }
}


$a = new fakeString();


echo $a->addA()->addB()->getStr();`

Comment (mickel1138) [reply src/Entity/DatasetSubmission.php]: How would you improve this particular scenario?

Comment (praneethpr) [reply src/Entity/DatasetSubmission.php]:
````
$newDatasetLink->setUrl($datasetLink->getUrl())->setName($datasetLink->getName())->setDescription($datasetLink->getDescription()) ......
````
We can do this, but I feel chaining methods like this is tough to understand

File: public/js/datasetReview.js [hunk 4]
```diff
@@ -114,6 +114,41 @@ $(document).ready(function(){
...
+        var value = parseFloat($(this).attr("index"));
+        datasetLinksCount = (value > datasetLinksCount) ? value : datasetLinksCount;
+    });
+
+    $("#addLink")
+        .button()
+        .click(function(){
+            datasetLinksCount++;
+
+            var newLink = $("#links-prototype table")
```
Comment (praneethpr) [inline public/js/datasetReview.js]: We should validate the Url using our API. Its done with the ERDDAP field

File: src/Form/DatasetLinkType.php [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+     * @return void
+     */
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $builder
+            ->add('url', Type\TextType::class, array(
+                'label' => 'Link Url:',
+                'required' => false,
+            ))
+            ->add('name', Type\TextType::class, array(
```
Comment (praneethpr) [inline src/Form/DatasetLinkType.php]: We can try using Datalist here, and the ticket describes this to be a predefined list.

File: src/Entity/DatasetLink.php (hunk 1 above)
Comment (fightingtexasaggie) [reply src/Entity/DatasetLink.php]: I didn't see your prior comment referred to, but would suggest checking that the past code is in the list of constants above.

File: src/Entity/DatasetSubmission.php [hunk 6]
```diff
@@ -1028,6 +1028,15 @@ class DatasetSubmission extends Entity
...
     protected $remotelyHostedFunction;
 
+    /**
+     * The Point of Contact for this Dataset Submission.
+     *
+     * @var Collection
+     *
+     * @ORM\OneToMany(targetEntity="App\Entity\DatasetLink", mappedBy="datasetSubmission", cascade={"persist"}, orphanRemoval=true)
+     */
+    private $datasetLinks;
```
Comment (fightingtexasaggie) [inline src/Entity/DatasetSubmission.php]: private vs protected?

File: src/Entity/DatasetSubmission.php (hunk 3 above)
Comment (fightingtexasaggie) [reply src/Entity/DatasetSubmission.php]: I'm actually ok with them too, but think we should decide one way or another and stick with it.

File: src/Entity/DatasetSubmission.php (hunk 6 above)
Comment (mickel1138) [reply src/Entity/DatasetSubmission.php]: This was added by the Maker bundle, but changed it.

File: templates/DatasetReview/dataset-links.html.twig [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+            </tr>
+        </table>
+        </fieldset>
+        {% endfor %}
+    </div>
+
+    <button id="addLink" type="button">Add Link</button>
+</fieldset>
+
+<datalist id="dataLinkNames">
```
Comment (praneethpr) [inline templates/DatasetReview/dataset-links.html.twig]: Should we also add Dataland ? It is mentioned in the ticket
I think this should be a list of constants in the DatasetLink Entity

File: src/Form/DatasetLinkType.php [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+     * @param array                $options The options.
+     *
+     * @see FormTypeExtensionInterface::buildForm()
+     *
+     * @return void
+     */
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $builder
+            ->add('url', Type\TextType::class, array(
```
Comment (praneethpr) [inline src/Form/DatasetLinkType.php]: It might be a question for the PO, Should these fields be required?

**Pull request merged**